atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/InvalidationRunner.java

```diff
@@ -0,0 +1,198 @@
+                        prefixedTimestampTableName()));
+            }
+        }
+    }
+
+    private Limits getLimits(Connection connection) throws SQLException {
+        DatabaseMetaData metaData = connection.getMetaData();
+        ResultSet res = metaData.getTables(null, null, prefixedTimestampTableName(), null);
+
+        Preconditions.checkState(res.next(), "We are in the process of invalidating the "
+                + "InDbTimestampBoundStore but the data table does not exist. "
+                + "We should never reach here. Please contact support.");
+
+        return ImmutableLimits.builder()
+                .upperLimit(getColumnStatus(LAST_ALLOCATED, connection))
+                .legacyUpperLimit(getColumnStatus(LEGACY_LAST_ALLOCATED, connection))
+                .build();
+    }
+
+    private ColumnStatus getColumnStatus(String colName, Connection connection) throws SQLException {
+        if (hasColumn(connection, colName)) {
+            String sql = String.format("SELECT %s FROM %s FOR UPDATE", colName, prefixedTimestampTableName());
+            QueryRunner run = new QueryRunner();
+            return run.query(connection, sql, rs -> {
+                if (rs.next()) {
+                    return ColumnStatus.columnStatusWithValue(rs.getLong(colName));
+                }
+                return ColumnStatus.columnStatusWithoutValue();
+            });
+
+        } else {
+            return ColumnStatus.voidColumnStatus();
+        }
+    }
+
+    private boolean hasColumn(Connection connection, String colName) throws SQLException {
+        return connection.getMetaData()
+                .getColumns(null, null, prefixedTimestampTableName(), colName)
+                .next();
+    }
+
+    private String prefixedTimestampTableName() {
+        return AtlasDbConstants.TIMELOCK_TIMESTAMP_TABLE.getQualifiedName();
+    }
+
+    private TableStatus checkTableStatus(Limits limits) {
+        TableStatus status = getTableStatus(limits);
+
+        Preconditions.checkState(status != TableStatus.ILLEGAL_COLUMNS,
+                "We detected the table has both current as well as legacy columns."
+                        + "This is unexpected. Please contact support.");
+        return status;
+    }
+
+    private TableStatus getTableStatus(Limits limits) {
+        boolean upperLimitExists = limits.upperLimit().exists();
+        boolean legacyUpperLimitExists = limits.legacyUpperLimit().exists();
+
+        if (upperLimitExists) {
+            return legacyUpperLimitExists ? TableStatus.ILLEGAL_COLUMNS : TableStatus.HEALTHY;
+        }
+        return legacyUpperLimitExists ? TableStatus.POISONED : TableStatus.NO_DATA; // no data in table
+    }
+
+    @Value.Immutable
+    interface Limits {
+        ColumnStatus upperLimit();
+        ColumnStatus legacyUpperLimit();
+    }
+
+    @Value.Immutable
+    interface ColumnStatus {
```

Comment:
I'd suggest having an `OptionalLong` for value here, and have users just use an `Optional<ColumnStatus>` rather than the null object pattern we're trying here